Cache prefixes of previously seen input when setting innerHTML

NathanSolomon733 opened the pull request:
#### cd3c9890ec057999b7580d14cc738c3b51ab0407
<pre>
Cache prefixes of previously seen input when setting innerHTML
<a href="https://bugs.webkit.org/show_bug.cgi?id=301231">https://bugs.webkit.org/show_bug.cgi?id=301231</a>
<a href="https://rdar.apple.com/162555307">rdar://162555307</a>

Reviewed by Ryosuke Niwa.

When a user sets innerHTML to an element, the next call to
innerHTML can often be just an extension of the previous markup.
Regardless, we reparse the entire input.
We can add a prefix match detection to tell the parser if the current
input to innerHTML is just an extension of a previous input.
If so, avoid jumping into the parser, deep clone the existing
tree, then just parse the unseen markup.
A prefix can also just be considered an exact match of a previous input.

This is 10% win on TODOMVC.Jquery.AddingAllItems.

Combined changes:
Tests: fast/parser/innerHTML-cache-mutation-test.html
       fast/parser/innerHTML-prefix-cache-boundary-test.html
* LayoutTests/fast/parser/innerHTML-cache-mutation-test-expected.txt: Added.
* LayoutTests/fast/parser/innerHTML-cache-mutation-test.html: Added.
* LayoutTests/fast/parser/innerHTML-prefix-cache-boundary-test-expected.txt: Added.
* LayoutTests/fast/parser/innerHTML-prefix-cache-boundary-test.html: Added.
* Source/WebCore/css/CSSStyleProperties.cpp:
(WebCore::InlineCSSStyleProperties::didMutate):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::parserNotifyChildrenChanged):
(WebCore::ContainerNode::childrenChanged):
(WebCore::ContainerNode::cloneSubtreeForFastParser const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::invalidateCachedSetInnerHTML):
(WebCore::Document::updateCachedSetInnerHTML):
* Source/WebCore/dom/Document.h:
(WebCore::Document::cachedSetInnerHTML):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::parseHTML):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::notifyAttributeChanged):
(WebCore::Element::classAttributeChanged):
(WebCore::Element::parserSetAttributes):
(WebCore::Element::replaceChildrenWithMarkup):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::invalidateNodeListCollectionAndInnerHTMLPrefixCachesInAncestorsForAttribute):
(WebCore::Node::setDidMutateSubtreeAfterSetInnerHTMLOnAncestors):
(WebCore::Node::setTextContent):
(WebCore::Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute): Deleted.
* Source/WebCore/dom/Node.h:
(WebCore::Node::hasDidMutateSubtreeAfterSetInnerHTML const):
(WebCore::Node::setDidMutateSubtreeAfterSetInnerHTML):
(WebCore::Node::clearDidMutateSubtreeAfterSetInnerHTML):
(WebCore::Node::hasWasParsedWithFastPath const):
(WebCore::Node::setWasParsedWithFastPath):
(WebCore::Node::clearWasParsedWithFastPath):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::replaceChildrenWithMarkup):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::inlineStyleChanged):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::maxCachedSetInnerHTMLStringSize):
(WebCore::isCachedSubtreeValid):
(WebCore::isCachedPrefixMatch):
(WebCore::tryAvoidParsingByCloningExistingSubtree):
(WebCore::cloneCachedPrefixAndParseSuffix):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::tryFastParsingHTMLFragmentImpl):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h:

Canonical link: <a href="https://commits.webkit.org/303089@main">https://commits.webkit.org/303089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80c717c43a90f03cd0a3bbcf068cce1ab1b9e7c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131232 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82956 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d896fd77-894b-44c7-ba47-4f9591f93700) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133102 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3404 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99990 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4861b37a-3f1f-41e8-a1b0-ce3e7c714fd9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134178 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2559 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80690 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c2af1c9b-2919-4ff9-a3a2-ca4cbd32f182) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2475 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/180 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81927 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111071 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35597 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141176 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3307 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36119 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108506 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2958 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108451 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27571 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2488 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113806 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56408 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3369 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32230 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3191 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66777 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3391 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3299 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -615,6 +626,7 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
     static auto flagIsHTML() { return enumToUnderlyingType(TypeFlag::IsHTMLElement); }
     static auto flagIsLink() { return enumToUnderlyingType(StateFlag::IsLink); }
     static auto flagIsParsingChildren() { return enumToUnderlyingType(StateFlag::IsParsingChildren); }
+    static auto flagDidNodeOrChildrenChange() { return enumToUnderlyingType(StateFlag::DidNodeOrChildrenChange); }
```
Surely, we're not updating this flag in JIT?

NathanSolomon733 replied:
This was a mistake and should not be added. I will remove it.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1102,6 +1108,24 @@ void Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute(const Qu
+        return;
+    // FIXME: could we early break once finding the cached containerNode?
+    for (auto* node = this; node; node = node->parentNode()) {
+        node->setDidNodeOrChildrenChange();
```
This flag name isn't accurate. We're setting this bit whenever any of the descendent node has been mutated.
I'd more directly call this something like `didMutateSubtreeAfterSetInnerHTML` so that the intent & semantics of the bit is crystal clear from its name.

NathanSolomon733 replied:
Agreed.

rniwa commented on Source/WebCore/dom/Element.h:
```diff
@@ -395,9 +395,12 @@ class Element : public ContainerNode {
+    Ref<Element> cloneElementWithChildren(Document&, CustomElementRegistry*, ContainerNode::ShouldNotifyChildInsertion = ContainerNode::ShouldNotifyChildInsertion::Yes) const;
     Ref<Element> cloneElementWithoutChildren(Document&, CustomElementRegistry*) const;
 
+    // Public wrapper for fast path parser to clone with skipped attribute notifications
```
Rather than adding a comment like this, we should give the function a more descriptive name:
e.g. `cloneSubtreeWithoutAttributeNotificationForFastPathParser`.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1053,17 +1071,20 @@ void ContainerNode::cloneChildNodes(Document& document, CustomElementRegistry* f
             clone.appendChildCommon(clonedChild);
-            notifyChildNodeInserted(clone, clonedChild, postInsertionNotificationTargets);
+            if (shouldNotify == ShouldNotifyChildInsertion::Yes)
+                notifyChildNodeInserted(clone, clonedChild, postInsertionNotificationTargets);
```
I can't imagine how this can be correct with respect to node flags that get updated upon insertion like node flags associated for effective lang.

NathanSolomon733 replied:
Oh I took a closer look and I see now that suppressing notifications can break effective lang inheritance since the cloned nodes wouldn't update based on their new parent. I thought we could skip this for faster parser cloning but you'd still have that problem. I'll remove this.

rniwa replied:
Yeah, I think there are a few other node flags that propagate values down the tree as we invoke insertedIntoAncestor / removedFromAncestor so I don't think can skip those without doing some refactoring first. It's doable but probably best to be tackled in a separate PR if any.

NathanSolomon733 commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1037,6 +1046,24 @@ void ContainerNode::childrenChanged(const ChildChange& change)
+        return;
+
+    for (RefPtr child = firstChild(); child; child = child->nextSibling()) {
+        Ref clonedChild = child->cloneNodeInternal(document, CloningOperation::SelfWithTemplateContent, fallbackRegistry);
```
I need to change the CloningOperation flag. This should be CloningOperation::SelfOnly. The fast path parser can't even parse template elements.

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -109,6 +109,12 @@ enum class TaskSource : uint8_t;
 using MutationObserverOptions = OptionSet<MutationObserverOptionType>;
 using MutationRecordDeliveryOptions = OptionSet<MutationObserverOptionType>;
 
+// Tracks the source of DOM mutations for cache invalidation
```
I don't think this comment adds much value. Remove?

NathanSolomon733 replied:
Will do.

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -109,6 +109,12 @@ enum class TaskSource : uint8_t;
 using MutationRecordDeliveryOptions = OptionSet<MutationObserverOptionType>;
 
+// Tracks the source of DOM mutations for cache invalidation
+enum class MutationSource : uint8_t {
```
I'd define `IsMutationBySetInnerHTML : bool { No, Yes }` instead since we don't have a third value yet.

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -674,7 +685,8 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
         InLargestContentfulPaintTextContentSet = 1 << 21,
-        // 11 bits free.
+        DidMutateSubtreeAfterSetInnerHTML = 1 << 22
+        // 10 bits free.
```
This label seems wrong? If we have 23 bits used for flags, we only have 9 bits left, not 10.

NathanSolomon733 replied:
Right. I will change that.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1082,18 +1082,32 @@ void Node::invalidateNodeListAndCollectionCachesInAncestors()
+        if (shouldSetMutationBit && !node->didMutateSubtreeAfterSetInnerHTML()) {
+            node->setDidMutateSubtreeAfterSetInnerHTML();
+            if (node == cachedContainer.get())
+                shouldSetMutationBit = false;
```
Is this right? Why is it okay to stop setting did mutate bit on ancestors? If a higher ancestor had a cache, that needs to be invalidated too, no?

NathanSolomon733 replied:
Currently there is just 1 cached pointer to a subtree root that is held as a member in Document. This pointer points to the subtree made for the largest markup string passed in as input into the fast parser. So, there is only one "cached" subtree per Document which makes it not needed to go all the way up the tree to set these mutation bits if we've already visited the 1 cachedContainer node.

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -95,6 +96,139 @@ class NewlineThenWhitespaceStringsTable {
+        return FragmentReuseResult::ParseFailure;
+    return FragmentReuseResult::Success;
+}
+void updateInnerHTMLPrefixCache(Document& document, const String& sourceString, ContainerNode& container, Element& contextElement)
```
Need a blank line before the function.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1037,6 +1046,24 @@ void ContainerNode::childrenChanged(const ChildChange& change)
 
 static constexpr size_t cloneMaxDepth = 1024;
 
+void ContainerNode::cloneChildNodesFastForParser(Document& document, CustomElementRegistry* fallbackRegistry, ContainerNode& clone, size_t currentDepth) const
```
I'd call this `cloneSubtreeForFastParser` instead.

rniwa commented on Source/WebCore/dom/DocumentFragment.cpp:
```diff
@@ -93,8 +93,10 @@ Ref<Node> DocumentFragment::cloneNodeInternal(Document& document, CloningOperati
 
 void DocumentFragment::parseHTML(const String& source, Element& contextElement, OptionSet<ParserContentPolicy> parserContentPolicy, CustomElementRegistry* registry)
 {
+    m_wasParsedWithFastPath = false;
```
I'd set this after the if statement for `tryFastParsingHTMLFragment` to avoid updating it twice for when m_wasParsedWithFastPath is true.

rniwa commented on Source/WebCore/dom/DocumentFragment.h:
```diff
@@ -52,6 +54,8 @@ class DocumentFragment : public ContainerNode {
     SerializedNode serializeNode(CloningOperation) const override;
     bool childTypeAllowed(NodeType) const override;
+
+    bool m_wasParsedWithFastPath { false };
```
This will grow the size of every `ShadowRoot`. That's not great.
I think we should use one of the remaining node flags to indicate this.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2319,7 +2320,10 @@ void Element::notifyAttributeChanged(const QualifiedName& name, const AtomString
     if (oldValue != newValue) {
-        invalidateNodeListAndCollectionCachesInAncestorsForAttribute(name);
+        MutationSource mutationSource = reason == AttributeModificationReason::InnerHTMLFastPath
+        ? MutationSource::InnerHTMLOnCachedContainer
```
Need to indent by 4 spaces.

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4386,7 +4390,12 @@ ExceptionOr<void> Element::replaceChildrenWithMarkup(const String& markup, Optio
+    bool usedFastPath = fragment.returnValue()->wasParsedWithFastPath();
+    auto result = replaceChildrenWithFragment(container, fragment.releaseReturnValue(), MutationSource::InnerHTMLOnCachedContainer);
+    if (!result.hasException() && usedFastPath)
+        updateInnerHTMLPrefixCache(document(), markup, container.get(), *this);
```
Please make this a member function of `Document`. There is no reason why it has to be a free standing function.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -541,10 +541,10 @@ ExceptionOr<void> Node::replaceChild(Node& newChild, Node& oldChild)
 }
 
-ExceptionOr<void> Node::removeChild(Node& oldChild)
+ExceptionOr<void> Node::removeChild(Node& oldChild, const MutationSource mutationSource)
```
DO we really need to specify `MutationSource` in this function?

NathanSolomon733 replied:
We do not. This is a mistake. I am removing it.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -713,12 +713,12 @@ ExceptionOr<void> Node::replaceWith(FixedVector<NodeOrString>&& nodeOrStringVect
 }
 
-ExceptionOr<void> Node::remove()
+ExceptionOr<void> Node::remove(const MutationSource mutationSource)
```
Ditto.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1082,18 +1082,32 @@ void Node::invalidateNodeListAndCollectionCachesInAncestors()
 }
 
-void Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute(const QualifiedName& attrName)
+void Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute(const QualifiedName& attrName, const MutationSource mutationSource)
```
We should rename this function now that we're also clearing the innerHTMLPrefix cache.
Maybe invalidateNodeListCollectionAndInnerHTMLPrefixCachesInAncestorsForAttribute

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1102,6 +1116,24 @@ void Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute(const Qu
     }
 }
 
+void Node::setDidMutateSubtreeAfterSetInnerHTML()
```
A trivial setter like this should be defined in `Node.h` instead.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1082,18 +1082,32 @@ void Node::invalidateNodeListAndCollectionCachesInAncestors()
 
     for (auto* node = this; node; node = node->parentNode()) {
+        if (shouldSetMutationBit && !node->didMutateSubtreeAfterSetInnerHTML()) {
+            node->setDidMutateSubtreeAfterSetInnerHTML();
```
I'd call this function just `didMutateSubtreeAfterSetInnerHTML` since that's the naming convention we use elsewhere.

NathanSolomon733 replied:
Are you suggesting I change the setter method name to `didMutateSubtreeAfterSetInnerHTML`? 
I was following the naming convention of `setHasHeldBackChildrenChanged`, which is why I named the setter with set*.
Did you mean to suggest changing the getter `didMutateSubtreeAfterSetInnerHTML` name to `hasDidMutateSubtreeAfterSetInnerHTML` to following the naming convention of `hasHeldBackChildrenChanged`? I realize that my getter is not following that convention.

rniwa commented on Source/WebCore/dom/Document.h:
```diff
@@ -382,6 +383,12 @@ enum class PageshowEventPersistence : bool { NotPersisted, Persisted };
 enum class EventHandlerRemoval : bool { One, All };
 using EventTargetSet = WeakHashCountedSet<Node, WeakPtrImplWithEventTargetData>;
 
+struct InnerHTMLPrefixCache {
```
This is a bit misleading since this isn't really a cache.
This structure is really storing the pointer to the subtree which could be simply cloned when setInnerHTML is called and the string with which the innerHTML was performed.

Maybe `CachedSetInnerHTML`?

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -95,6 +96,139 @@ class NewlineThenWhitespaceStringsTable {
+}
+
+template<typename CharacterType>
+static FragmentReuseResult tryReuseCachedFragment(std::span<const CharacterType> source, ContainerNode& destinationParent, Element& contextElement)
```
This isn't really reusing "cache" since the cache in this case is really a DOM subtree.
I think it's better to call it something like `tryAvoidParsingByCloningExistingSubtree`.

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -95,6 +96,139 @@ class NewlineThenWhitespaceStringsTable {
+static FragmentReuseResult tryReuseCachedFragment(std::span<const CharacterType> source, ContainerNode& destinationParent, Element& contextElement)
+{
+    // FIXME: implement caching for the parsed results of small markup strings
+    if (source.size() < s_minCachedInputSize)
```
Why do we need this lower limit at all?

NathanSolomon733 replied:
We don't need it, I just thought the added overhead of this optimization could make it not useful for small inputs, where I doubted fast parsing a small markup string is any slower than deep cloning it. This decision was not based in any tests or data.

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -95,6 +96,139 @@ class NewlineThenWhitespaceStringsTable {
+        return FragmentReuseResult::CannotReuse;
+
+    auto unparsedSuffix = source.subspan(cache.source.length());
+    return cloneCachedPrefixAndParseSuffix(unparsedSuffix, destinationParent, contextElement, cachedContainer.get());
```
Hm... is this really going to always work? Is it true that the fast parser's result is always additive?
Because that's certainly not the case for general HTML parsing.

NathanSolomon733 replied:
It is only true for a very specific subtest TODOMVC.Jquery.AddingAllItems.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1037,6 +1046,24 @@ void ContainerNode::childrenChanged(const ChildChange& change)
+
+    for (RefPtr child = firstChild(); child; child = child->nextSibling()) {
+        Ref clonedChild = child->cloneNodeInternal(document, CloningOperation::SelfOnly, fallbackRegistry);
+        executeParserNodeInsertionIntoIsolatedTreeWithoutNotifyingParent(clone, clonedChild.get(), [&] {
```
We need to call childrenChanged at some point for each parent.
It's not correct to never call `childrenChanged`.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1037,6 +1039,25 @@ void ContainerNode::childrenChanged(const ChildChange& change)
+        if (RefPtr childAsContainerNode = dynamicDowncast<ContainerNode>(*child))
+            childAsContainerNode->cloneSubtreeForFastParser(document, fallbackRegistry, downcast<ContainerNode>(clonedChild.get()), currentDepth + 1);
+    }
+    if (clone.hasHeldBackChildrenChanged())
```
I don't think we need to check this condition. `executeParserNodeInsertionIntoIsolatedTreeWithoutNotifyingParent` will definitely holdback `childrenChanged`. We should probably just assert that.

NathanSolomon733 replied:
Oh, right. I will remove it.
Also, I checked and ContainerNode::parserNotifyChildrenChanged() already contains an ASSERT(hasHeldBackChildrenChanged()) check at the beginning of the function, so that is ASSERT is already done.

NathanSolomon733 replied:
I realized it was possible to call `parserNotifyChildrenChanged` without the `setHasHeldBackChildrenChanged` flag set first if we called `cloneSubtreeForFastParser` on a void element like `<br>` that is technically a ContainerNode but has no children. The loop over its children would not execute because the `<br>` has no children. `setHasHeldBackChildrenChanged` would never get called, but then `parserNotifyChildrenChanged` still gets called at the end of the function even though no children were added.

 To fix this, I added checks to ensure `cloneSubtreeForFastParser` is only called on nodes that have children.

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -1082,26 +1082,52 @@ void Node::invalidateNodeListAndCollectionCachesInAncestors()
-        if (!node->hasRareData())
+        if (shouldSetMutationBit && !node->hasDidMutateSubtreeAfterSetInnerHTML()) {
+            node->setDidMutateSubtreeAfterSetInnerHTML();
+            if (node == cachedContainer.get())
```
Don't we want to `invalidateCachedSetInnerHTML` here?

NathanSolomon733 replied:
We don't need to eagerly because `invalidateCachedSetInnerHTML` will be called on the next fast parse in `tryAvoidParsingByCloningExistingSubtree` if the cached container gets setDidMutateSubtreeAfterSetInnerHTML called on it here. That function checks if this Node flag has been set.

rniwa commented on Source/WebCore/dom/StyledElement.cpp:
```diff
@@ -216,6 +216,7 @@ RefPtr<StyleProperties> StyledElement::protectedInlineStyle() const
 
 void StyledElement::inlineStyleChanged()
 {
+    // FIXME: do we need a call to setDidMutateSubtreeAfterSetInnerHTMLOnAncestors() here?
```
We most certainly should. Add a test which does `element.style.display = 'none'` or something.

NathanSolomon733 replied:
Added.

NathanSolomon733 committed:
Cache prefixes of previously seen input when setting innerHTML
https://bugs.webkit.org/show_bug.cgi?id=301231
rdar://162555307

Reviewed by Ryosuke Niwa.

When a user sets innerHTML to an element, the next call to
innerHTML can often be just an extension of the previous markup.
Regardless, we reparse the entire input.
We can add a prefix match detection to tell the parser if the current
input to innerHTML is just an extension of a previous input.
If so, avoid jumping into the parser, deep clone the existing
tree, then just parse the unseen markup.
A prefix can also just be considered an exact match of a previous input.

This is 10% win on TODOMVC.Jquery.AddingAllItems.

Combined changes:
Tests: fast/parser/innerHTML-cache-mutation-test.html
       fast/parser/innerHTML-prefix-cache-boundary-test.html
* LayoutTests/fast/parser/innerHTML-cache-mutation-test-expected.txt: Added.
* LayoutTests/fast/parser/innerHTML-cache-mutation-test.html: Added.
* LayoutTests/fast/parser/innerHTML-prefix-cache-boundary-test-expected.txt: Added.
* LayoutTests/fast/parser/innerHTML-prefix-cache-boundary-test.html: Added.
* Source/WebCore/css/CSSStyleProperties.cpp:
(WebCore::InlineCSSStyleProperties::didMutate):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::parserNotifyChildrenChanged):
(WebCore::ContainerNode::childrenChanged):
(WebCore::ContainerNode::cloneSubtreeForFastParser const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::invalidateCachedSetInnerHTML):
(WebCore::Document::updateCachedSetInnerHTML):
* Source/WebCore/dom/Document.h:
(WebCore::Document::cachedSetInnerHTML):
* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::parseHTML):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::notifyAttributeChanged):
(WebCore::Element::classAttributeChanged):
(WebCore::Element::parserSetAttributes):
(WebCore::Element::replaceChildrenWithMarkup):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::invalidateNodeListCollectionAndInnerHTMLPrefixCachesInAncestorsForAttribute):
(WebCore::Node::setDidMutateSubtreeAfterSetInnerHTMLOnAncestors):
(WebCore::Node::setTextContent):
(WebCore::Node::invalidateNodeListAndCollectionCachesInAncestorsForAttribute): Deleted.
* Source/WebCore/dom/Node.h:
(WebCore::Node::hasDidMutateSubtreeAfterSetInnerHTML const):
(WebCore::Node::setDidMutateSubtreeAfterSetInnerHTML):
(WebCore::Node::clearDidMutateSubtreeAfterSetInnerHTML):
(WebCore::Node::hasWasParsedWithFastPath const):
(WebCore::Node::setWasParsedWithFastPath):
(WebCore::Node::clearWasParsedWithFastPath):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::replaceChildrenWithMarkup):
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::inlineStyleChanged):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::maxCachedSetInnerHTMLStringSize):
(WebCore::isCachedSubtreeValid):
(WebCore::isCachedPrefixMatch):
(WebCore::tryAvoidParsingByCloningExistingSubtree):
(WebCore::cloneCachedPrefixAndParseSuffix):
(WebCore::HTMLFastPathParser::parseAttributes):
(WebCore::tryFastParsingHTMLFragmentImpl):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.h:

Canonical link: https://commits.webkit.org/303089@main
